planner/core/logical_plan_builder.go

```diff
@@ -2421,6 +2421,95 @@ func (b *PlanBuilder) resolveCorrelatedAggregates(ctx context.Context, sel *ast.
 	return correlatedAggMap, nil
 }
 
+type asofResolver struct {
+	ctx      sessionctx.Context
+	tsValues []*types.Datum
+	err      error
+}
+
+func (a *asofResolver) Enter(inNode ast.Node) (ast.Node, bool) {
+	switch n := inNode.(type) {
+	case *ast.AsOfClause:
+		if n.TsExpr != nil {
+			tsRes, err := evalAstExpr(a.ctx, n.TsExpr)
+			if err != nil {
+				a.err = err
+				return n, true
+			}
+			a.tsValues = append(a.tsValues, &tsRes)
+
+		}
+		return n, false
+	// for the cases: select * from a as of timestamp 'xxx', b;
+	case *ast.TableName:
+		if n.AsOf == nil {
+			a.tsValues = append(a.tsValues, nil)
+		}
+		return n, false
+	}
+
+	return inNode, false
+}
+
+func (a *asofResolver) Leave(inNode ast.Node) (ast.Node, bool) {
+	return inNode, true
+}
+
+// TryExtractTSFromAsOf trys to extract the specified timestamp.
+func TryExtractTSFromAsOf(ctx sessionctx.Context, node ast.Node) (*types.Datum, error) {
+	switch x := node.(type) {
+	case *ast.SelectStmt:
+		tblRefs := x.From
+		if tblRefs != nil {
+			resolve := &asofResolver{
+				ctx: ctx,
+			}
+			tblRefs.Accept(resolve)
+			if resolve.err != nil {
+				return nil, resolve.err
+			}
+			tsValues := resolve.tsValues
+			if len(tsValues) == 0 {
+				return nil, nil
+			}
+
+			var res *types.Datum
+			first := tsValues[0]
+			if first != nil {
+				ts, err := first.ConvertTo(ctx.GetSessionVars().StmtCtx, types.NewFieldType(mysql.TypeTimestamp))
+				if err != nil {
+					return nil, err
+				}
+				res = &ts
+			}
+
+			for i := 1; i < len(tsValues); i++ {
```

Comment:
This check should be a global check?
`select ... from (select ... from t as of ...) tmp_1, (select ... from t2 as of ...) tmp_2 ...`
We should guarantee that the AsOfs in the two sub selects use the same ts?

Reply:
Also for another case, `select ... from (select ... from t as of ...) tmp_1 as of xxx` is valid or not?

Reply:
Yes, it should be a global check. If `startTS` is same, then it is valid. Because we bind only one txn with one `startTS`.

It is kinda meaningless talking about `tmp_1 as of xxx`, the temporary created from sub select stmt does not have a history.